--- cacti-1.2.9/include/config.php.orig	2020-02-10 02:25:45.000000000 +0100
+++ cacti-1.2.9/include/config.php	2020-02-10 10:46:35.436772588 +0100
@@ -124,7 +124,7 @@ $disable_log_rotation = false;
  * the CRSF secret file.
  */
 
-//$path_csrf_secret = '/usr/share/cacti/resource/csrf-secret.php';
+$path_csrf_secret = '/var/lib/cacti/csrf/csrf-secret.php';
 
 /*
  * The following are optional variables for debugging low level system
