--- a/library/Director/Objects/IcingaObjectGroups.php
+++ b/library/Director/Objects/IcingaObjectGroups.php
@@ -173,7 +173,7 @@
             }
 
             $this->groups[$group->object_name] = $group;
-        } elseif (is_string($group)) {
+        } elseif (is_string($group) || is_integer($group)) {
             if (array_key_exists($group, $this->groups)) {
                 return $this;
             }
--- a/library/vendor/ipl/Web/Table/QueryBasedTable.php
+++ b/library/vendor/ipl/Web/Table/QueryBasedTable.php
@@ -110,7 +110,7 @@
 
     public function renderContent()
     {
-        if (count($this->getColumnsToBeRendered())) {
+        if ($this->getColumnsToBeRendered() && count($this->getColumnsToBeRendered())) {
             $this->generateHeader();
         }
         $this->fetchRows();
